### PR TITLE
LESS and SASS Sourcemaps and a couple bugfixes

### PR DESCRIPTION
**Pull request by DeMoorJasper**
# ↪️ Pull Request

This PR implements sourcemaps for LESS and SASS, it also fixes some bugs with sourcemaps I discover along the way.

## ✔️ PR Todo

- [x] Added/updated unit tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 6.90s | +56.00ms |
| Cached | 2.44s | +21.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 389.00ms | +131.00ms ⚠️ |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 389.00ms | +131.00ms ⚠️ |
| dist/legacy/kitchen-sink.32bb4226.js | 1.13kb | +0.00b | 999.00ms | -699.00ms 🚀 |
| dist/modern/kitchen-sink.280a0cce.js | 1.13kb | +0.00b | 675.00ms | -1.02s 🚀 |
| dist/legacy/index.html | 709.00b | +0.00b | 907.00ms | +196.00ms ⚠️ |
| dist/modern/index.html | 709.00b | +0.00b | 903.00ms | +194.00ms ⚠️ |
| dist/legacy/styles.afb8e31a.css | 78.00b | +0.00b | 1.59s | -108.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 200.00ms | -39.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 197.00ms | -42.00ms 🚀 |
| dist/legacy/kitchen-sink.32bb4226.js | 1.13kb | +0.00b | 487.00ms | -52.00ms 🚀 |
| dist/modern/kitchen-sink.280a0cce.js | 1.13kb | +0.00b | 484.00ms | -59.00ms 🚀 |
| dist/legacy/index.html | 709.00b | +0.00b | 488.00ms | -53.00ms 🚀 |
| dist/modern/index.html | 709.00b | +0.00b | 484.00ms | -54.00ms 🚀 |
| dist/legacy/styles.afb8e31a.css | 78.00b | +0.00b | 488.00ms | -62.00ms 🚀 |
| dist/modern/styles.cda8c91b.css | 78.00b | +0.00b | 484.00ms | -65.00ms 🚀 |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 10.56s | +83.00ms |
| Cached | 2.45s | -30.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/main/index.js | 66.80kb | +0.00b | 223.00ms | -73.00ms 🚀 |
| dist/module/index.js | 37.45kb | +0.00b | 214.00ms | -81.00ms 🚀 |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 3.26m | -3.00s |
| Cached | 5.02s | +74.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/EmojiPickerComponent.487f0492.js | 98.73kb | +0.00b | 30.50s | +2.85s ⚠️ |
| dist/component.6c698016.js | 48.66kb | +0.00b | 11.72s | +1.08s ⚠️ |
| dist/Modal.f79bc3a9.js | 42.03kb | +0.00b | 6.37s | -625.00ms 🚀 |
| dist/component.2d909511.js | 40.07kb | +0.00b | 12.55s | +1.98s ⚠️ |
| dist/clipboard.03c807f0.js | 17.21kb | +0.00b | 1.44m | +26.29s ⚠️ |
| dist/js.786dad09.js | 16.49kb | +0.00b | 6.35s | -610.00ms 🚀 |
| dist/16.d3236fb9.js | 10.22kb | +0.00b | 6.36s | -596.00ms 🚀 |
| dist/16.c42426b9.js | 10.09kb | +0.00b | 6.35s | -609.00ms 🚀 |
| dist/16.bc89f6ee.js | 10.02kb | +0.00b | 6.35s | -609.00ms 🚀 |
| dist/16.98f3baa0.js | 9.98kb | +0.00b | 12.41s | +1.92s ⚠️ |
| dist/fallback.aab3a799.js | 8.90kb | +0.00b | 19.09s | +955.00ms ⚠️ |
| dist/code.6b50d9e2.js | 8.69kb | +0.00b | 15.66s | +3.17s ⚠️ |
| dist/heading5.6844d092.js | 8.48kb | +0.00b | 20.86s | +1.28s ⚠️ |
| dist/action.536bde1c.js | 8.31kb | +0.00b | 13.76s | +1.26s ⚠️ |
| dist/panel-warning.cb245541.js | 8.29kb | +0.00b | 1.95m | +1.67m ⚠️ |
| dist/list.624b775a.js | 8.26kb | +0.00b | 16.88s | -1.62m 🚀 |
| dist/16.c4ee6c00.js | 5.17kb | +0.00b | 11.42s | +1.27s ⚠️ |
| dist/16.7f86346a.js | 4.89kb | +0.00b | 12.31s | +1.21s ⚠️ |
| dist/16.5f13de00.js | 4.84kb | +0.00b | 11.03s | +881.00ms ⚠️ |
| dist/16.51c4a0ba.js | 4.82kb | +0.00b | 14.96s | +2.46s ⚠️ |
| dist/16.c81c0695.js | 4.82kb | +0.00b | 14.96s | +2.46s ⚠️ |
| dist/16.61fcf150.js | 4.78kb | +0.00b | 14.96s | +3.85s ⚠️ |
| dist/16.232fb9da.js | 4.69kb | +0.00b | 13.63s | +1.13s ⚠️ |
| dist/16.60895c0e.js | 4.65kb | +0.00b | 12.56s | +1.46s ⚠️ |
| dist/16.771934f3.js | 4.65kb | +0.00b | 13.16s | +3.01s ⚠️ |
| dist/16.8a21dead.js | 4.64kb | +0.00b | 12.31s | +1.21s ⚠️ |
| dist/16.ba7ed30e.js | 4.64kb | +0.00b | 12.31s | +1.21s ⚠️ |
| dist/16.ea6e7d4b.js | 4.63kb | +0.00b | 11.03s | +881.00ms ⚠️ |
| dist/ak-editor.47799eae.css | 3.43kb | +0.00b | 7.86s | -1.40m 🚀 |
| dist/ResourcedEmojiComponent.4495d16e.js | 2.32kb | +0.00b | 12.41s | +1.92s ⚠️ |
| dist/media-card-analytics-error-boundary.030a93f8.js | 1.62kb | +0.00b | 44.16s | +3.80s ⚠️ |
| dist/index.html | 212.00b | +0.00b | 1.53m | +1.42m ⚠️ |

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>Three.js x4 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/262864235586912769)**

File: packages/core/integration-tests/test/sourcemaps.js [hunk 1]
```diff
@@ -638,14 +638,14 @@ describe('sourcemaps', function() {
...
-  it.skip('should create a valid sourcemap for a SASS asset', async function() {
+  it('should create a valid sourcemap for a SASS asset', async function() {
     async function test(minify) {
       let inputFilePath = path.join(
         __dirname,
         '/integration/sourcemap-sass/style.scss',
       );
 
-      await bundle(inputFilePath, {minify});
+      await bundle(inputFilePath, {minify, patchConsole: false});
```
Comment (mischnic) [inline packages/core/integration-tests/test/sourcemaps.js:648]: Was this intentional?

File: packages/transformers/sass/src/SassTransformer.js [hunk 2]
```diff
@@ -83,6 +94,17 @@ export default new Transformer({
           asset.addIncludedFile({filePath: included});
         }
       }
+
+      if (result.map != null) {
+        let map = new SourceMap();
+        let {mappings, sources, names} = JSON.parse(result.map);
+        map.addRawMappings(
+          mappings,
+          sources, // .map(s => path.relative(options.projectRoot, s)),
```
Comment (mischnic) [inline packages/transformers/sass/src/SassTransformer.js:103]: Should that be commented back in?

**Comment by mischnic** [comment]
While you're at it, can you pin the sourcemap package to the exact version?
According to semver, `2.0.0-nightly.123` is actually newer than `2.0.0-alpha.4.9`. It works with Yarn 1 but Yarn 2 uses the old nightly version of the sourcemap.

File: packages/core/integration-tests/test/sourcemaps.js (hunk 1 above)
Comment (DeMoorJasper) [reply packages/core/integration-tests/test/sourcemaps.js:648]: Nope it's debug, will remove it

File: packages/transformers/sass/src/SassTransformer.js (hunk 2 above)
Comment (DeMoorJasper) [reply packages/transformers/sass/src/SassTransformer.js:103]: nope, was just copy-paste from less but apparently sass already does relative mapping

Comment (mischnic) [reply packages/transformers/sass/src/SassTransformer.js]: How does it know it should be relative to the projectRoot?

Comment (DeMoorJasper) [reply packages/transformers/sass/src/SassTransformer.js]: > `OutFile`: This string option is the location that Sass expects the generated CSS to be saved to. It’s used to determine the URL used to link from the generated CSS to the source map, and from the source map to the Sass source files.

It's due to `outFile` will remove it and see if that outputs the full paths

**Pull request merged**